File: apps/mobile-parent/lib/subscription/coverage_profile_service.dart
/// Coverage Profile Service for Parent App
///
/// Fetches and caches coverage profiles from the billing service.
/// Provides methods to check feature access and display coverage info.
library;

import 'dart:async';
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import 'coverage_profile_models.dart';

// ══════════════════════════════════════════════════════════════════════════════
// CONFIGURATION
// ══════════════════════════════════════════════════════════════════════════════

const _billingBaseUrl = String.fromEnvironment(
  'BILLING_BASE_URL',
  defaultValue: 'http://localhost:4060',
);

const _useMock = bool.fromEnvironment('USE_COVERAGE_MOCK', defaultValue: true);

// Cache duration
const _cacheDuration = Duration(minutes: 5);

// ══════════════════════════════════════════════════════════════════════════════
// SERVICE
// ══════════════════════════════════════════════════════════════════════════════

/// Service for fetching and managing coverage profiles.
class CoverageProfileService {
  CoverageProfileService({String? accessToken}) {
    final headers = accessToken != null 
        ? {'Authorization': 'Bearer $accessToken'} 
        : <String, dynamic>{};
    
    _dio = Dio(BaseOptions(baseUrl: _billingBaseUrl, headers: headers));
  }

  late final Dio _dio;

  // Cache
  final Map<String, _CachedProfile> _cache = {};

  // ──────────────────────────────────────────────────────────────────────────
  // COVERAGE PROFILE OPERATIONS
  // ──────────────────────────────────────────────────────────────────────────

  /// Get coverage profile for a learner.
  Future<CoverageProfile> getCoverageProfile({
    required String learnerId,
    required String tenantId,
    required int grade,
    String? schoolId,
    bool forceRefresh = false,
  }) async {
    final cacheKey = 'profile:$learnerId';
    
    // Check cache
    if (!forceRefresh) {
      final cached = _cache[cacheKey];
      if (cached != null && cached.isValid) {
        return cached.profile;
      }
    }

    // Fetch from service
    CoverageProfile profile;
    if (_useMock) {
      profile = _mockCoverageProfile(learnerId, tenantId, grade, schoolId);
    } else {
      final response = await _dio.get(
        '/api/v1/coverage/$learnerId',
        queryParameters: {
          'tenantId': tenantId,
          'grade': grade,
          if (schoolId != null) 'schoolId': schoolId,
          'includeDetails': true,
          'forceRefresh': forceRefresh,
        },
      );
      
      final data = response.data as Map<String, dynamic>;
      profile = CoverageProfile.fromJson(data['profile'] as Map<String, dynamic>);
    }

    // Update cache
    _cache[cacheKey] = _CachedProfile(
      profile: profile,
      expiresAt: DateTime.now().add(_cacheDuration),
    );

    return profile;
  }

  /// Get coverage profile summary (lightweight).
  Future<CoverageProfileSummary> getCoverageProfileSummary({
    required String learnerId,
    required String tenantId,
    required int grade,
    String? schoolId,
  }) async {
    // Get full profile (from cache if available)
    final profile = await getCoverageProfile(
      learnerId: learnerId,
      tenantId: tenantId,
      grade: grade,
      schoolId: schoolId,
    );

    return CoverageProfileSummary(
      learnerId: profile.learnerId,
      hasDistrictBase: profile.hasDistrictCoverage,
      hasParentSubscription: profile.hasParentCoverage,
      districtFeatureCount: profile.districtModules.length,
      parentFeatureCount: profile.parentModules.length,
      totalEffectiveFeatures: profile.effectiveModules.length,
      refundableOverlapCount: profile.parentCoverage?.overlappingFeatures.length ?? 0,
    );
  }

  /// Check if a learner has access to a specific feature.
  Future<FeatureAccessResult> checkFeatureAccess({
    required String learnerId,
    required String featureKey,
    required String tenantId,
    required int grade,
    String? schoolId,
  }) async {
    if (_useMock) {
      return _mockFeatureAccess(featureKey);
    }

    final response = await _dio.get(
      '/api/v1/coverage/$learnerId/feature/$featureKey',
      queryParameters: {
        'tenantId': tenantId,
        'grade': grade,
        if (schoolId != null) 'schoolId': schoolId,
      },
    );

    final data = response.data as Map<String, dynamic>;
    return FeatureAccessResult(
      hasAccess: data['hasAccess'] as bool,
      providedBy: FeaturePayer.fromCode(data['providedBy'] as String),
      displayLabel: data['displayLabel'] as String,
    );
  }

  /// Get batch coverage profiles for multiple learners.
  Future<Map<String, CoverageProfile>> getBatchCoverageProfiles({
    required List<LearnerInfo> learners,
    bool includeDetails = false,
  }) async {
    if (_useMock) {
      return {
        for (final l in learners)
          l.learnerId: _mockCoverageProfile(l.learnerId, l.tenantId, l.grade, l.schoolId),
      };
    }

    final response = await _dio.post(
      '/api/v1/coverage/batch',
      data: {
        'learners': learners.map((l) => <String, dynamic>{
            'learnerId': l.learnerId,
            'tenantId': l.tenantId,
            'schoolId': l.schoolId,
            'grade': l.grade,
          }).toList(),
        'includeDetails': includeDetails,
      },
    );

    final data = response.data as Map<String, dynamic>;
    final profiles = data['profiles'] as Map<String, dynamic>;
    
    return profiles.map(
      (k, v) => MapEntry(k, CoverageProfile.fromJson(v as Map<String, dynamic>)),
    );
  }

  /// Invalidate cached profile for a learner.
  void invalidateProfile(String learnerId) {
    _cache.remove('profile:$learnerId');
  }

  /// Clear all cached profiles.
  void clearCache() {
    _cache.clear();
  }

  // ──────────────────────────────────────────────────────────────────────────
  // MOCK DATA
  // ──────────────────────────────────────────────────────────────────────────

  CoverageProfile _mockCoverageProfile(
    String learnerId,
    String tenantId,
    int grade,
    String? schoolId,
  ) {
    // Simulate a learner with both district and parent coverage
    final districtModules = <String>{'MODULE_ELA', 'MODULE_MATH'};
    final parentModules = <String>{'ADDON_SEL', 'FEATURE_HOMEWORK_HELPER'};
    final overlapping = <String>{}; // No overlap in mock
    
    final allModules = {...districtModules, ...parentModules};
    
    return CoverageProfile(
      learnerId: learnerId,
      grade: grade,
      tenantId: tenantId,
      schoolId: schoolId,
      hasDistrictCoverage: true,
      hasParentCoverage: true,
      districtCoverage: DistrictCoverage(
        tenantId: tenantId,
        contractId: 'mock-contract-id',
        contractNumber: 'DST-2025-001',
        schoolId: schoolId,
        coveredFeatures: districtModules,
        startDate: DateTime(2025, 1, 1),
        endDate: DateTime(2026, 6, 30),
        isActive: true,
      ),
      parentCoverage: ParentCoverage(
        billingAccountId: 'mock-billing-account',
        subscriptionId: 'mock-subscription-id',
        linkedLearnerId: learnerId,
        coveredFeatures: parentModules,
        overlappingFeatures: overlapping,
        periodStart: DateTime.now().subtract(const Duration(days: 15)),
        periodEnd: DateTime.now().add(const Duration(days: 15)),
        isActive: true,
        status: 'ACTIVE',
      ),
      districtModules: districtModules,
      parentModules: parentModules,
      effectiveModules: allModules,
      payerForFeature: {
        'MODULE_ELA': FeaturePayer.district,
        'MODULE_MATH': FeaturePayer.district,
        'ADDON_SEL': FeaturePayer.parent,
        'FEATURE_HOMEWORK_HELPER': FeaturePayer.parent,
      },
      coverageDetails: [
        const FeatureCoverageDetail(
          featureKey: 'MODULE_ELA',
          payer: FeaturePayer.district,
          sourceId: 'mock-contract-id',
          displayLabel: 'Provided by your school',
          expiresAt: null,
        ),
        const FeatureCoverageDetail(
          featureKey: 'MODULE_MATH',
          payer: FeaturePayer.district,
          sourceId: 'mock-contract-id',
          displayLabel: 'Provided by your school',
          expiresAt: null,
        ),
        FeatureCoverageDetail(
          featureKey: 'ADDON_SEL',
          payer: FeaturePayer.parent,
          sourceId: 'mock-subscription-id',
          displayLabel: 'Your subscription',
          expiresAt: DateTime.now().add(const Duration(days: 15)),
        ),
        FeatureCoverageDetail(
          featureKey: 'FEATURE_HOMEWORK_HELPER',
          payer: FeaturePayer.parent,
          sourceId: 'mock-subscription-id',
          displayLabel: 'Your subscription',
          expiresAt: DateTime.now().add(const Duration(days: 15)),
        ),
      ],
      upsellOpportunities: [
        'MODULE_SCIENCE',
        'ADDON_SPEECH',
        'ADDON_TUTORING',
      ],
      computedAt: DateTime.now(),
      expiresAt: DateTime.now().add(const Duration(minutes: 5)),
    );
  }

  FeatureAccessResult _mockFeatureAccess(String featureKey) {
    // Mock: district provides ELA/MATH, parent provides SEL
    const districtFeatures = {'MODULE_ELA', 'MODULE_MATH'};
    const parentFeatures = {'ADDON_SEL', 'FEATURE_HOMEWORK_HELPER'};

    if (districtFeatures.contains(featureKey)) {
      return const FeatureAccessResult(
        hasAccess: true,
        providedBy: FeaturePayer.district,
        displayLabel: 'Provided by your school',
      );
    }
    if (parentFeatures.contains(featureKey)) {
      return const FeatureAccessResult(
        hasAccess: true,
        providedBy: FeaturePayer.parent,
        displayLabel: 'Your subscription',
      );
    }
    return const FeatureAccessResult(
      hasAccess: false,
      providedBy: FeaturePayer.none,
      displayLabel: 'Available for upgrade',
    );
  }
}

// ══════════════════════════════════════════════════════════════════════════════
// HELPER CLASSES
// ══════════════════════════════════════════════════════════════════════════════

/// Cached profile with expiration.
class _CachedProfile {
  _CachedProfile({required this.profile, required this.expiresAt});

  final CoverageProfile profile;
  final DateTime expiresAt;

  bool get isValid => DateTime.now().isBefore(expiresAt);
}

/// Result of feature access check.
class FeatureAccessResult {
  const FeatureAccessResult({
    required this.hasAccess,
    required this.providedBy,
    required this.displayLabel,
  });

  final bool hasAccess;
  final FeaturePayer providedBy;
  final String displayLabel;
}

/// Learner info for batch requests.
class LearnerInfo {
  const LearnerInfo({
    required this.learnerId,
    required this.tenantId,
    this.schoolId,
    required this.grade,
  });

  final String learnerId;
  final String tenantId;
  final String? schoolId;
  final int grade;
}

// ══════════════════════════════════════════════════════════════════════════════
// PROVIDER
// ══════════════════════════════════════════════════════════════════════════════

/// Provider for coverage profile service.
final coverageProfileServiceProvider = Provider<CoverageProfileService>((ref) {
  // In real app, get access token from auth state
  // final accessToken = ref.watch(authStateProvider).accessToken;
  return CoverageProfileService();
});

/// Provider for a specific learner's coverage profile.
final coverageProfileProvider = FutureProvider.family<CoverageProfile, LearnerInfo>(
  (ref, learnerInfo) async {
    final service = ref.watch(coverageProfileServiceProvider);
    return service.getCoverageProfile(
      learnerId: learnerInfo.learnerId,
      tenantId: learnerInfo.tenantId,
      grade: learnerInfo.grade,
      schoolId: learnerInfo.schoolId,
    );
  },
);

/// Provider for checking feature access.
final featureAccessProvider = FutureProvider.family<FeatureAccessResult, FeatureAccessParams>(
  (ref, params) async {
    final service = ref.watch(coverageProfileServiceProvider);
    return service.checkFeatureAccess(
      learnerId: params.learnerId,
      featureKey: params.featureKey,
      tenantId: params.tenantId,
      grade: params.grade,
      schoolId: params.schoolId,
    );
  },
);

/// Parameters for feature access check.
class FeatureAccessParams {
  const FeatureAccessParams({
    required this.learnerId,
    required this.featureKey,
    required this.tenantId,
    required this.grade,
    this.schoolId,
  });

  final String learnerId;
  final String featureKey;
  final String tenantId;
  final int grade;
  final String? schoolId;

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is FeatureAccessParams &&
        other.learnerId == learnerId &&
        other.featureKey == featureKey &&
        other.tenantId == tenantId &&
        other.grade == grade &&
        other.schoolId == schoolId;
  }

  @override
  int get hashCode {
    return Object.hash(learnerId, featureKey, tenantId, grade, schoolId);
  }
}

File: services/billing-svc/src/services/coverage-profile.service.ts
/**
 * Coverage Profile Service
 *
 * Computes and manages learner coverage profiles that combine district
 * entitlements with parent subscriptions.
 *
 * PRECEDENCE RULES (implemented here):
 * 1. District coverage always wins for overlapping features
 * 2. Parent subscriptions only cover features NOT provided by district
 * 3. Features not covered by either result in upsell opportunity
 */

import {
  coverageProfileRepository,
  type DistrictEntitlementData,
  type LearnerInfo,
  type ParentSubscriptionData,
} from '../repositories/coverage-profile.repository.js';
import {
  type CoverageProfile,
  type CoverageProfileSummary,
  type CoverageSourceType,
  CoverageSourceTypeValues,
  type DistrictCoverage,
  type FeatureCoverageDetail,
  type FeatureKey,
  FeatureKeyValues,
  type FeaturePayer,
  FeaturePayerValues,
  type ParentCoverage,
  computeEffectiveModules,
  computeFeaturePayer,
  findOverlappingFeatures,
  profileToSummary,
  type ParentAddonAggregateView,
  type AddonBreakdownItem,
  type TopAddonItem,
} from '../types/coverage-profile.types.js';
import { gradeToGradeBand, type GradeBand } from '../types/licensing.types.js';

// ============================================================================
// Cache Configuration
// ============================================================================

const CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes
const profileCache = new Map<string, { profile: CoverageProfile; expiresAt: number }>();
// ============================================================================
// Coverage Profile Service
// ============================================================================

export class CoverageProfileService {
  /**
   * Compute coverage profile for a learner.
   * This is the main entry point for determining what features a learner has
   * and who is paying for them.
   */
  async getCoverageProfile(
    learnerId: string,
    learnerInfo: LearnerInfo,
    forceRefresh = false
  ): Promise<CoverageProfile> {
    const cacheKey = `profile:${learnerId}`;

    // Check cache
    if (!forceRefresh) {
      const cached = profileCache.get(cacheKey);
      if (cached && cached.expiresAt > Date.now()) {
        return cached.profile;
      }
    }

    // Compute fresh profile
    const profile = await this.computeCoverageProfile(learnerInfo);

    // Cache result
    profileCache.set(cacheKey, {
      profile,
      expiresAt: Date.now() + CACHE_TTL_MS,
    });

    return profile;
  }

  /**
   * Get coverage profile summary for UI display.
   */
  async getCoverageProfileSummary(
    learnerId: string,
    learnerInfo: LearnerInfo
  ): Promise<CoverageProfileSummary> {
    const profile = await this.getCoverageProfile(learnerId, learnerInfo);
    return profileToSummary(profile);
  }

  /**
   * Batch compute coverage profiles for multiple learners.
   */
  async getBatchCoverageProfiles(
    learnerInfos: LearnerInfo[]
  ): Promise<Map<string, CoverageProfile>> {
    const results = new Map<string, CoverageProfile>();

    // Process in parallel with concurrency limit
    const concurrencyLimit = 10;
    for (let i = 0; i < learnerInfos.length; i += concurrencyLimit) {
      const batch = learnerInfos.slice(i, i + concurrencyLimit);
      const profiles = await Promise.all(
        batch.map((info) => this.getCoverageProfile(info.learnerId, info))
      );
      batch.forEach((info, idx) => {
        results.set(info.learnerId, profiles[idx]);
      });
    }

    return results;
  }

  /**
   * Clear cached profile for a learner.
   * Call this when coverage changes (e.g., subscription change, contract update).
   */
  invalidateProfile(learnerId: string): void {
    profileCache.delete(`profile:${learnerId}`);
  }

  /**
   * Clear all cached profiles for a tenant.
   * Call this when district contract changes affect multiple learners.
   */
  invalidateTenantProfiles(_tenantId: string): void {
    // In a real implementation, you'd track learner->tenant mapping
    // For now, clear entire cache when tenant changes
    profileCache.clear();
  }

  // --------------------------------------------------------------------------
  // Core Computation
  // --------------------------------------------------------------------------

  /**
   * Core computation of coverage profile.
   * Implements precedence rules: district wins for overlaps.
   */
  private async computeCoverageProfile(learnerInfo: LearnerInfo): Promise<CoverageProfile> {
    const { learnerId, tenantId, schoolId, grade } = learnerInfo;
    const gradeBand = gradeToGradeBand(grade);

    // Fetch district entitlements and parent subscription in parallel
    const [districtEntitlements, parentSubscription] = await Promise.all([
      coverageProfileRepository.getDistrictEntitlements(tenantId, schoolId),
      coverageProfileRepository.getParentSubscription(learnerId),
    ]);

    // Build district coverage
    const districtCoverage = this.buildDistrictCoverage(
      tenantId,
      schoolId,
      gradeBand,
      districtEntitlements
    );

    // Build parent coverage (with overlap detection)
    const districtModules = districtCoverage?.coveredFeatures ?? new Set<string>();
    const parentCoverage = this.buildParentCoverage(learnerId, parentSubscription, districtModules);

    // Compute effective modules
    const parentModules = parentCoverage?.coveredFeatures ?? new Set<string>();
    const effectiveModules = computeEffectiveModules(districtModules, parentModules);

    // Build payer map with precedence rules
    const payerForFeature = new Map<string, FeaturePayer>();
    for (const feature of effectiveModules) {
      payerForFeature.set(feature, computeFeaturePayer(feature, districtModules, parentModules));
    }

    // Build coverage details
    const coverageDetails = this.buildCoverageDetails(
      districtCoverage,
      parentCoverage,
      payerForFeature
    );

    // Determine upsell opportunities (features not covered)
    const upsellOpportunities = Object.values(FeatureKeyValues).filter(
      (feature) => !effectiveModules.has(feature) && !districtModules.has(feature)
    );

    const now = new Date();
    return {
      learnerId,
      grade,
      gradeBand,
      tenantId,
      schoolId,
      hasDistrictCoverage: districtCoverage !== null && districtModules.size > 0,
      hasParentCoverage: parentCoverage !== null && parentModules.size > 0,
      districtCoverage,
      parentCoverage,
      districtModules,
      parentModules,
      effectiveModules,
      payerForFeature,
      coverageDetails,
      upsellOpportunities,
      computedAt: now,
      expiresAt: new Date(now.getTime() + CACHE_TTL_MS),
    };
  }

  /**
   * Build district coverage from entitlements.
   */
  private buildDistrictCoverage(
    tenantId: string,
    schoolId: string | null,
    gradeBand: GradeBand,
    entitlements: DistrictEntitlementData[]
  ): DistrictCoverage | null {
    if (entitlements.length === 0) {
      return null;
    }

    // Group by contract (typically there's one active contract)
    const byContract = new Map<string, DistrictEntitlementData[]>();
    for (const ent of entitlements) {
      const existing = byContract.get(ent.contractId) ?? [];
      existing.push(ent);
      byContract.set(ent.contractId, existing);
    }

    // Use the first active contract (could be extended to merge multiple)
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const [contractId, contractEntitlements] = byContract.entries().next().value!;
    const firstEntitlement = contractEntitlements[0];

    // Collect all feature keys from entitlements
    const coveredFeatures = new Set<string>();
    let minStartDate = firstEntitlement.startDate;
    let maxEndDate = firstEntitlement.endDate;

    for (const ent of contractEntitlements) {
      coveredFeatures.add(ent.featureKey);
      if (ent.startDate < minStartDate) minStartDate = ent.startDate;
      if (ent.endDate > maxEndDate) maxEndDate = ent.endDate;
    }

    return {
      tenantId,
      contractId,
      contractNumber: firstEntitlement.contractNumber,
      schoolId,
      gradeBand,
      coveredFeatures,
      startDate: minStartDate,
      endDate: maxEndDate,
      isActive: true,
    };
  }

  /**
   * Build parent coverage from subscription.
   * Detects overlaps with district coverage.
   */
  private buildParentCoverage(
    learnerId: string,
    subscription: ParentSubscriptionData | null,
    districtModules: Set<string>
  ): ParentCoverage | null {
    if (!subscription) {
      return null;
    }

    const coveredFeatures = new Set(subscription.coveredFeatures);
    const overlappingFeatures = findOverlappingFeatures(districtModules, coveredFeatures);

    return {
      billingAccountId: subscription.billingAccountId,
      subscriptionId: subscription.subscriptionId,
      linkedLearnerId: learnerId,
      coveredFeatures,
      overlappingFeatures,
      periodStart: subscription.periodStart,
      periodEnd: subscription.periodEnd,
      isActive: subscription.status === 'ACTIVE' || subscription.status === 'IN_TRIAL',
      status: subscription.status,
    };
  }

  /**
   * Build detailed coverage info for each feature.
   */
  private buildCoverageDetails(
    districtCoverage: DistrictCoverage | null,
    parentCoverage: ParentCoverage | null,
    payerForFeature: Map<string, FeaturePayer>
  ): FeatureCoverageDetail[] {
    const details: FeatureCoverageDetail[] = [];

    for (const [featureKey, payer] of payerForFeature) {
      let sourceType: CoverageSourceType | null = null;
      let sourceId: string | null = null;
      let displayLabel: string;
      let expiresAt: Date | null = null;

      if (payer === FeaturePayerValues.DISTRICT && districtCoverage) {
        sourceType = CoverageSourceTypeValues.DISTRICT_CONTRACT;
        sourceId = districtCoverage.contractId;
        displayLabel = 'Provided by your school';
        expiresAt = districtCoverage.endDate;
      } else if (payer === FeaturePayerValues.PARENT && parentCoverage) {
        sourceType = featureKey.startsWith('ADDON_')
          ? CoverageSourceTypeValues.PARENT_ADDON_SUBSCRIPTION
          : CoverageSourceTypeValues.PARENT_BASE_SUBSCRIPTION;
        sourceId = parentCoverage.subscriptionId;
        displayLabel = 'Your subscription';
        expiresAt = parentCoverage.periodEnd;
      } else {
        displayLabel = 'Not covered';
      }

      details.push({
        featureKey,
        payer,
        sourceType,
        sourceId,
        displayLabel,
        expiresAt,
      });
    }

    return details;
  }

  // --------------------------------------------------------------------------
  // District Admin Views
  // --------------------------------------------------------------------------

  /**
   * Get aggregate view of parent add-ons for district admin dashboard.
   * Shows what add-ons parents are purchasing for learners in the district.
   */
  async getParentAddonAggregateView(
    tenantId: string,
    schoolId?: string | null
  ): Promise<ParentAddonAggregateView> {
    // Get district entitlements
    const districtEntitlements = await coverageProfileRepository.getDistrictEntitlements(
      tenantId,
      schoolId
    );
    const districtFeatures = new Set(districtEntitlements.map((e) => e.featureKey));

    // Get parent subscriptions
    const parentSubscriptions = await coverageProfileRepository.getParentSubscriptionsForTenant(
      tenantId,
      schoolId
    );

    // Count learners (simplified - would come from learner service)
    const totalLearners = 1000; // Placeholder - integrate with learner service
    const learnersWithParentSubs = new Set(
      parentSubscriptions.map((s) => s.linkedLearnerId).filter(Boolean)
    ).size;

    // Build addon breakdown
    const featureCounts: Record<FeatureKey, { district: number; parent: number }> = {} as Record<
      FeatureKey,
      { district: number; parent: number }
    >;
    const allFeatures = Object.values(FeatureKeyValues);

    for (const feature of allFeatures) {
      featureCounts[feature] = {
        district: districtFeatures.has(feature) ? totalLearners : 0,
        parent: 0,
      };
    }

    // Count parent purchases per feature
    for (const sub of parentSubscriptions) {
      for (const feature of sub.coveredFeatures) {
        // Only count features that are in our known feature list
        if (Object.prototype.hasOwnProperty.call(featureCounts, feature)) {
          featureCounts[feature as FeatureKey].parent++;
        }
      }
    }

    const addonBreakdown: AddonBreakdownItem[] = allFeatures.map((feature) => ({
      featureKey: feature,
      displayName: getFeatureDisplayName(feature),
      districtCoveredCount: featureCounts[feature].district,
      parentPurchasedCount: featureCounts[feature].parent,
      notCoveredCount: Math.max(
        0,
        totalLearners - featureCounts[feature].district - featureCounts[feature].parent
      ),
    }));

    // Find top parent add-ons (opportunity for district to bundle)
    const topParentAddons: TopAddonItem[] = addonBreakdown
      .filter((a) => a.parentPurchasedCount > 0)
      .sort((a, b) => b.parentPurchasedCount - a.parentPurchasedCount)
      .slice(0, 5)
      .map((a) => ({
        featureKey: a.featureKey,
        displayName: a.displayName,
        purchaseCount: a.parentPurchasedCount,
        // Rough estimate: $5/month * count * 12 months
        potentialSavingsEstimateCents: a.parentPurchasedCount * 500 * 12,
      }));

    return {
      tenantId,
      schoolId: schoolId ?? null,
      totalLearners,
      learnersWithParentSubs,
      parentCoveragePercentage: (learnersWithParentSubs / totalLearners) * 100,
      addonBreakdown,
      topParentAddons,
      computedAt: new Date(),
    };
  }

  // --------------------------------------------------------------------------
  // Feature Access Checks
  // --------------------------------------------------------------------------

  /**
   * Check if a learner has access to a specific feature.
   */
  async hasFeatureAccess(
    learnerId: string,
    learnerInfo: LearnerInfo,
    featureKey: string
  ): Promise<boolean> {
    const profile = await this.getCoverageProfile(learnerId, learnerInfo);
    return profile.effectiveModules.has(featureKey);
  }

  /**
   * Get the payer for a specific feature.
   */
  async getFeaturePayer(
    learnerId: string,
    learnerInfo: LearnerInfo,
    featureKey: string
  ): Promise<FeaturePayer> {
    const profile = await this.getCoverageProfile(learnerId, learnerInfo);
    return profile.payerForFeature.get(featureKey) ?? FeaturePayerValues.NONE;
  }

  /**
   * Check if a feature is provided by district (for UI "Provided by school" label).
   */
  async isFeatureProvidedByDistrict(
    learnerId: string,
    learnerInfo: LearnerInfo,
    featureKey: string
  ): Promise<boolean> {
    const payer = await this.getFeaturePayer(learnerId, learnerInfo, featureKey);
    return payer === FeaturePayerValues.DISTRICT;
  }
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Get human-readable display name for a feature key.
 */
function getFeatureDisplayName(featureKey: string): string {
  const displayNames: Record<string, string> = {
    [FeatureKeyValues.MODULE_ELA]: 'English Language Arts',
    [FeatureKeyValues.MODULE_MATH]: 'Mathematics',
    [FeatureKeyValues.MODULE_SCIENCE]: 'Science',
    [FeatureKeyValues.ADDON_SEL]: 'Social-Emotional Learning',
    [FeatureKeyValues.ADDON_SPEECH]: 'Speech & Language',
    [FeatureKeyValues.ADDON_TUTORING]: 'AI Tutoring',
    [FeatureKeyValues.FEATURE_HOMEWORK_HELPER]: 'Homework Helper',
    [FeatureKeyValues.FEATURE_PROGRESS_REPORTS]: 'Progress Reports',
    [FeatureKeyValues.FEATURE_PARENT_INSIGHTS]: 'Parent Insights',
  };
  return displayNames[featureKey] ?? featureKey;
}

// Export singleton instance
export const coverageProfileService = new CoverageProfileService();
